cache_flat_mutation_reader.hh

```diff
@@ -730,9 +749,39 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
         }
     }
 
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
     if (_next_row_in_range) {
+        bool remove_row = false;
+
+        if (_read_context.tombstone_gc_state() // do not compact rows when tombstone_gc_state is not set (used in some unit tests)
+            && !_next_row.dummy()
+            && _snp->at_latest_version()
+            && _snp->at_oldest_version()) {
+            deletable_row& row = _next_row.latest_row();
+            auto t = row.deleted_at();
+
+            auto row_tomb_expired = [&](row_tombstone tomb) {
+                return (tomb && tomb.max_deletion_time() < _gc_before);
+            };
+
+            auto is_row_dead = [&](const deletable_row& row) {
+                return (!row.marker().is_missing() && row.marker().is_dead(_read_time));
+            };
+
+            if (row_tomb_expired(t) || is_row_dead(row)) {
```

Comment:
do you mean check `row.marker().deletion_time() < _gc_before`?

Reply:
Yes.

Reply:
well, but do we really need `.is_dead()` check after this condition?

Reply:
We do, because we don't want to interpret deletion_time() on live cells. It's also illegal to call deletion_time() on just any marker:

```
// Should be called when is_dead() or is_expiring().
```